### PR TITLE
Reimpl meaningful test name lint MCP658

### PR DESCRIPTION
**Pull request by workingjubilee**
This reintroduces the tidy rule originally proposed in https://github.com/rust-lang/rust/pull/113583 that then became an MCP in https://github.com/rust-lang/compiler-team/issues/658 which eventually surfaced a quite-reasonable request for a diagnostic enhancement. I have added that to the rule. It produces output like this:
```
tidy error: file `ui/unsized/issue-115809.rs` must begin with a descriptive name, try `{reason}-issue-115809.rs`
tidy error: file `ui/unsized/issue-115203.rs` must begin with a descriptive name, try `{reason}-issue-115203.rs`
tidy error: file `ui/privacy/issue-113860-2.rs` must begin with a descriptive name, try `{reason}-issue-113860.rs`
tidy error: file `ui/privacy/issue-117997.rs` must begin with a descriptive name, try `{reason}-issue-117997.rs`
tidy error: file `ui/privacy/issue-119463.rs` must begin with a descriptive name, try `{reason}-issue-119463.rs`
tidy error: file `ui/privacy/auxiliary/issue-117997.rs` must begin with a descriptive name, try `{reason}-issue-117997.rs`
tidy error: file `ui/privacy/auxiliary/issue-119463-extern.rs` must begin with a descriptive name, try `{reason}-issue-119463.rs`
tidy error: file `ui/privacy/issue-113860-1.rs` must begin with a descriptive name, try `{reason}-issue-113860.rs`
tidy error: file `ui/privacy/issue-113860.rs` must begin with a descriptive name, try `{reason}-issue-113860.rs`
tidy error: file `ui/const-generics/generic_const_exprs/const_kind_expr/issue_114151.rs` must begin with a descriptive name, try `{reason}-issue-114151.rs`
tidy error: file `ui/did_you_mean/issue-114112.rs` must begin with a descriptive name, try `{reason}-issue-114112.rs`
tidy error: file `ui/did_you_mean/issue-105225.rs` must begin with a descriptive name, try `{reason}-issue-105225.rs`
tidy error: file `ui/did_you_mean/issue-105225-named-args.rs` must begin with a descriptive name, try `{reason}-issue-105225.rs`
```

You get the idea.

There are some tests which merely would require reordering of the name according to the rule. I could modify the diagnostic further to identify those, but doing such would make it prone to bad suggestions. I have opted to trust contributors to recognize the diagnostic is robotic, as the pattern we are linting on is easier to match if we do not speculate on what parts of the name are meaningful: sometimes a word is a reason, but sometimes it is a mere "tag", such as with a pair like:
- issue-314159265-blue.rs
- issue-314159265-red.rs

Starting them with `red-` and `blue-` means they do not sort together, despite being related, and the color names are still not very descriptive. Recognizing a good name is an open-ended task, though this pair might be:
- colored-circle-gen-blue.rs
- colored-circle-gen-red.rs

Deciding exactly *how* to solve this is not the business of tidy, only recognizing a what.

r? @compiler-errors

### CODE REVIEW

**Comment by workingjubilee** [comment]
...of course.

**Comment by Nilstrieb** [comment]
@bors p=20
let's not interfere with the current rollups and get them through, but after that and another rebase of yours this really needs some p

File: src/tools/tidy/src/issues.txt [hunk 1]
```diff
@@ -0,0 +1,4376 @@
+/*
```
Comment (Nilstrieb) [inline src/tools/tidy/src/issues.txt:1]: ~4000/~15000.... that's a lot of grandfathers 🫠

File: src/tools/tidy/src/ui_tests.rs [hunk 2]
```diff
@@ -135,6 +144,38 @@ pub fn check(path: &Path, bad: &mut bool) {
...
         }
     });
+
+    // if an excluded file is renamed, it must be removed from this list
+    if allowed_issue_filenames.len() > 0 {
+        for file_name in allowed_issue_filenames {
+            let mut p = PathBuf::from(path);
+            p.push(file_name);
+            tidy_error!(
+                bad,
```
Comment (Nilstrieb) [inline src/tools/tidy/src/ui_tests.rs:175]: it would be very cool if tidy `--bless`ing removed it for you

**Comment by compiler-errors** [comment]
I'm in favor of this change, but given its size, do you mind if we wait a few days for the bors queue to calm down a bit?

File: src/tools/tidy/src/ui_tests.rs (hunk 2 above)
Comment (workingjubilee) [reply src/tools/tidy/src/ui_tests.rs:175]: Hmm, that sounds plausible.

**Comment by workingjubilee** [comment]
No problem if you help me remember when / give me a poke to rebase it by the usual venues.

**Comment by compiler-errors** [comment]
howdy @workingjubilee, I think I'm ready to review this if you're able to rebase it. I'd also really like if `./x.py test tidy --bless` did what nils asked for above!

@rustbot author

File: src/tools/tidy/src/issues.txt [hunk 3]
```diff
@@ -390,26 +391,29 @@
 "ui/borrowck/issue-95079-missing-move-in-nested-closure.rs",
 "ui/box/issue-82446.rs",
 "ui/box/issue-95036.rs",
+"ui/c-variadic/issue-32201.rs",
+"ui/c-variadic/issue-86053-1.rs",
+"ui/c-variadic/issue-86053-2.rs",
```
Comment (workingjubilee) [inline src/tools/tidy/src/issues.txt:396]: The new order was not quite stable for this round, but it should be in the future, as BTreeSet yields items in order when iterated and uses the Ord of String.

**Comment by petrochenkov** [comment]
@workingjubilee 
On Windows issues.txt is always regenerated with backslashes in paths, instead of forward slashes like in the committed version of the file.

**Comment by workingjubilee** [comment]
>  On Windows issues.txt is always regenerated with backslashes in paths, instead of forward slashes like in the committed version of the file.

Odd! Noted and will fix, apologies.

**Pull request merged**